### PR TITLE
refactor(core): bake single-thread assumptions into spawn/spawn_blocking

### PR DESCRIPTION
**Pull request by mmastrac**
Partially supersedes #19016.

This migrates `spawn` and `spawn_blocking` to `deno_core`, and removes the requirement for `spawn` tasks to be `Send` given our single-threaded executor.

While we don't need to technically do anything w/`spawn_blocking`, this allows us to have a single `JoinHandle` type that works for both cases, and allows us to more easily experiment with alternative `spawn_blocking` implementations that do not require tokio (ie: rayon).

Async ops (+~35%):

Before: 

```
time 1310 ms rate 763358
time 1267 ms rate 789265
time 1259 ms rate 794281
time 1266 ms rate 789889
```

After:

```
time 956 ms rate 1046025
time 954 ms rate 1048218
time 924 ms rate 1082251
time 920 ms rate 1086956
```

HTTP serve (+~4.4%):

Before:

```
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    68.78us   19.77us   1.43ms   86.84%
    Req/Sec    68.78k     5.00k   73.84k    91.58%
  1381833 requests in 10.10s, 167.36MB read
Requests/sec: 136823.29
Transfer/sec:     16.57MB
```

After:

```
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    63.12us   17.43us   1.11ms   85.13%
    Req/Sec    71.82k     3.71k   77.02k    79.21%
  1443195 requests in 10.10s, 174.79MB read
Requests/sec: 142921.99
Transfer/sec:     17.31MB
```

Suggested-By: alice@ryhl.io
Co-authored-by: Bartek Iwańczuk <biwanczuk@gmail.com>

### CODE REVIEW

**Review by bartlomieju — COMMENTED** [review]
Something to consider for a follow up - add `clippy.toml` that disallows using `tokio::task::spawn` and other methods and suggests to use `deno_core::task:*`

File: cli/tests/integration/cert_tests.rs [hunk 1]
```diff
@@ -219,113 +218,99 @@ fn cafile_bundle_remote_exports() {
 
 #[tokio::test]
 async fn listen_tls_alpn() {
-  // TLS streams require the presence of an ambient local task set to gracefully
-  // close dropped connections in the background.
```
Comment (bartlomieju) [inline cli/tests/integration/cert_tests.rs:223]: Not anymore 😎

File: cli/tools/task.rs [hunk 2]
```diff
@@ -59,9 +60,10 @@ pub async fn execute_script(
     let seq_list = deno_task_shell::parser::parse(&script)
       .with_context(|| format!("Error parsing script '{task_name}'."))?;
     let env_vars = collect_env_vars();
-    let exit_code =
-      deno_task_shell::execute(seq_list, env_vars, &cwd, Default::default())
-        .await;
+    let local = LocalSet::new();
+    let future =
+      deno_task_shell::execute(seq_list, env_vars, &cwd, Default::default());
+    let exit_code = local.run_until(future).await;
```
Comment (bartlomieju) [inline cli/tools/task.rs:66]: Why is `LocalSet` required here? Is it coming from `deno_task_shell` typings? If so I will update `deno_task_shell` to not require it if possible (CC @dsherret)

File: core/task.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+/// Equivalent to [`tokio::task::JoinHandle`].
+#[repr(transparent)]
+pub struct JoinHandle<R> {
+  handle: tokio::task::JoinHandle<MaskResultAsSend<R>>,
+  _r: PhantomData<R>,
+}
+
+impl<R> JoinHandle<R> {
+  /// Equivalent to [`tokio::task::JoinHandle::abort`].
+  pub fn abort(&self) {
```
Comment (bartlomieju) [inline core/task.rs:19]: Do we ever use it? In LSP maybe?

File: core/task.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+/// Equivalent to [`tokio::task::spawn`], but does not require the future to be [`Send`]. Must only be
+/// used on a [`RuntimeFlavor::CurrentThread`] executor, though this is only checked when running with
+/// debug assertions.
+pub fn spawn<F: Future<Output = R> + 'static, R: 'static>(
+  f: F,
+) -> JoinHandle<R> {
+  debug_assert!(
+    Handle::current().runtime_flavor() == RuntimeFlavor::CurrentThread
+  );
```
Comment (bartlomieju) [inline core/task.rs:52]: Nice!

**Review by bartlomieju — APPROVED** [review]
This is great, I love that you didn't have to mark structs in `serde_v8` as implementing `Send`.

LGTM 🚀

File: cli/tools/task.rs (hunk 2 above)
Comment (dsherret) [reply cli/tools/task.rs:66]: https://github.com/denoland/deno_task_shell/blob/f70223aa98593751dd83b3397fbdc469145b4850/src/shell/mod.rs#LL118C41-L118C52

Comment (bartlomieju) [reply cli/tools/task.rs:66]: Actually, thinking about it - it's probably not that bad that we use `LocalSet` for this specific purpose.

Comment (mmastrac) [reply cli/tools/task.rs:66]: I think we should pass an Executor in there.

File: core/task.rs (hunk 3 above)
Comment (mmastrac) [reply core/task.rs:19]: It was needed in a few places, but that was the only API required.

File: cli/cache/cache_db.rs [hunk 5]
```diff
@@ -95,7 +96,7 @@ impl Drop for CacheDB {
       // Hand off SQLite connection to another thread to do the surprisingly expensive cleanup
       let inner = inner.into_inner().into_inner();
       if let Some(conn) = inner {
-        tokio::task::spawn_blocking(move || {
```
Comment (dsherret) [inline cli/cache/cache_db.rs:98]: Maybe we should add some clippy rules banning the use of tokio::task::spawn and spawn_blocking? (I think by using `disallowed_methods`, but not sure)

Comment (bartlomieju) [reply cli/cache/cache_db.rs:98]: https://github.com/denoland/deno/pull/19056#pullrequestreview-1419095438

**Comment by bartlomieju** [comment]
Let's not land it yet, it appears this PR makes the problem in https://github.com/denoland/deno/issues/19100 even worse (ie. it takes more time to respond that that interval call). Let's investigate that issue a bit before we land, maybe it's not related, but over 1ms overhead to clamped 4ms interval seems like a serious bug.

**Comment by bartlomieju** [comment]
After further discussion offline we decided to land this PR and address the timeout problems in a follow up.

**Pull request merged**